### PR TITLE
Drop libselinux dependency from libsystemd

### PR DESCRIPTION
**Pull request by keszybz**
My original goal was to make libselinux optional in the initramfs by using dlopen() for it.
But I started by moving code that refers to libselinux from src/basic/ to src/shared/.
This turns out be useful in itself, because the linkage list for libsystemd.so is pruned.
Using dlopen() for libselinux is still possible, but I'll try to submit that separately, becuase
it's less clear that it is worth the trouble.

### CODE REVIEW

**Comment by keszybz** [comment]
It seems the total size is also reduced:
```console
$ ls -Ll {../old/,}build/libsystemd.so
-rwxrwxr-x 1 zbyszek zbyszek 5584264 Jun 23 10:52 ../old/build/libsystemd.so*
-rwxrwxr-x 1 zbyszek zbyszek 5323496 Jun 23 10:47 build/libsystemd.so*
$ ls -Ll {../old/,}build/src/shared/libsystemd-shared-249.so
-rwxrwxr-x 1 zbyszek zbyszek 11800784 Jun 23 10:52 ../old/build/src/shared/libsystemd-shared-249.so*
-rwxrwxr-x 1 zbyszek zbyszek 11789752 Jun 23 10:47 build/src/shared/libsystemd-shared-249.so*
$ du -sh /var/tmp/inst.{old,}
249M	/var/tmp/inst.old
248M	/var/tmp/inst
```
(-O2, no stripping)

**Comment by keszybz** [comment]
(In case it wasn't clear from the original description: there is no functional change. This is just moving code around a bit to move the libselinux dependency from the low-level library to higher level components that actually make use of it.)

**Comment by doverride** [comment]
Which means that I can remove the excessive libselinux related accesses from my security policy once this trickles down. Pity that this had to trickle down in the first place. Oh well i guess it keeps me off the streets.

**Comment by keszybz** [comment]
> Which means that I can remove the excessive libselinux related accesses

I didn't really verify this, but I think there are no changes in how the selinux code is called. Essentially, the code wasn't used by anything in libsystemd, so apart from reading `/usr/lib/libselinux.so.0.something` and the tiny bit of extra resources spent on this, there should be no change.

**Comment by doverride** [comment]
Okay, I guess I will see. Since a release or two many apps that have no business with SELinux started to do libselinux operations. The only thing they *seem* to have in common is that they link with libsystemd either directly or indirectly (nss_systemd).

**Comment by keszybz** [comment]
Dunno. In principle the library could have a shared global constructor to invoke some code on initialization… I don't know how to check if libselinux has any though.

**Comment by doverride** [comment]
Does not matter, once this trickles down I will remove the libselinux related accesses from some domains that I know have no business with SELinux. That should then clarify whether this was it or something else.

**Comment by yuwata** [comment]
LGTM.

**Comment by yuwata** [comment]
Do you want to this to be included in v249? or later?

File: src/shared/tpm2-util.c [hunk 1]
```diff
@@ -42,84 +42,39 @@ TSS2_RC (*sym_Tss2_MU_TPM2B_PUBLIC_Marshal)(TPM2B_PUBLIC const *src, uint8_t buf
...
-                k++;
-        }
+        r = dlopen_many_sym_or_warn(
+                        &libtss2_rc_dl, "libtss2-rc.so.0", LOG_DEBUG,
+                        DLSYM_ARG(Tss2_RC_Decode));
+        if (r < 0)
+                return r;
 
-        return k;
+        r = dlopen_many_sym_or_warn(
```
Comment (bluca) [inline src/shared/tpm2-util.c]: `return dlopen_many_sym_or_warn(` ?

**Review by bluca — APPROVED** [review]
LGTM, just one minor nit

**Comment by poettering** [comment]
I wonder if the libselinux' gcc constructor idiocy will fuck with us here. i.e. they do complicated setup from a gcc constructor, and we'll possibly delay running that quite a bit, this way, no? previously we ran this at process start for most cases, but now it will be done on dlopen() time. Which hopefully isn't an issue, but might need some careful checking, after all we run things during earlies boot where careful scheduling of the steps matters since so little is set up yet.

File: src/basic/mountpoint-util.c [hunk 2]
```diff
@@ -510,25 +509,3 @@ int mount_propagation_flags_from_string(const char *name, unsigned long *ret) {
                 return -EINVAL;
         return 0;
 }
-
-int make_mount_point_inode_from_stat(const struct stat *st, const char *dest, mode_t mode) {
-        assert(st);
-        assert(dest);
-
-        if (S_ISDIR(st->st_mode))
-                return mkdir_label(dest, mode);
```
Comment (poettering) [inline src/basic/mountpoint-util.c:519]: how did this even compile?

File: src/basic/fd-util.c [hunk 3]
```diff
@@ -476,6 +479,17 @@ int fd_get_path(int fd, char **ret) {
...
+int fd_is_on_btrfs(int fd) {
+        struct statfs sfs;
+
+        assert(fd >= 0);
+
+        if (fstatfs(fd, &sfs) < 0)
+                return -errno;
+
+        return F_TYPE_EQUAL(sfs.f_type, BTRFS_SUPER_MAGIC);
+}
```
Comment (poettering) [inline src/basic/fd-util.c]: drop the function altogether.  `fd_is_fs_type(fd, BTRFS_SUPER_MAGIC)` does the same and is not much longer to type

File: src/basic/fd-util.c [hunk 4]
```diff
@@ -1057,3 +1071,20 @@ int read_nr_open(void) {
         /* If we fail, fall back to the hard-coded kernel limit of 1024 * 1024. */
         return 1024 * 1024;
 }
+
+/* This is here because it's fd-related, and is called from sd-journal code, and we don't want to pull in
+ * shared/ there. */
```
Comment (poettering) [inline src/basic/fd-util.c]: why not? can you explain?

File: src/basic/copy.h [hunk 5]
```diff
@@ -57,6 +57,9 @@ static inline int copy_directory(const char *from, const char *to, CopyFlags cop
         return copy_directory_full(from, to, copy_flags, NULL, NULL, NULL);
 }
 
+int acquire_data_fd(const void *data, size_t size, unsigned flags);
+int copy_data_fd(int fd);
```
Comment (poettering) [inline src/basic/copy.h]: hmm, do they really fit into copy.c? 

maybe just give them their own .c/.h file? they are complex enough for that. i.e. `data-fd-util.[ch]`?

File: src/shared/dlfcn-util.c [hunk 6]
```diff
@@ -2,39 +2,63 @@
...
+        int r;
+
+        va_start(ap, log_level);
+        r = dlsym_many_or_warnv(dl, log_level, ap);
+        va_end(ap);
+
+        return r;
+}
+
+int dlopen_many_sym_or_warn_sentinel(void **ret_dl, const char *filename, int log_level, ...) {
```
Comment (poettering) [inline src/shared/dlfcn-util.c]: `ret_` is supposed to be the prefix for return variables, i.e. output params. but this one is both input *and* output. maybe name dlp?

File: src/basic/mountpoint-util.c (hunk 2 above)
Comment (keszybz) [reply src/basic/mountpoint-util.c:519]: You mean if selinux is disabled? The `#if HAVE_SELINUX` conditionals are in selinux-util.c, internally in the functions. W/o selinux the functions are noops, but they are still declared.

File: src/basic/fd-util.c (hunk 4 above)
Comment (keszybz) [reply src/basic/fd-util.c]: General policy. Thanks to @bluca, we even have this documented: https://systemd.io/ARCHITECTURE/#shared-code. I'll rework the comment to make this more explicit.

File: src/basic/copy.h (hunk 5 above)
Comment (keszybz) [reply src/basic/copy.h]: OK, will do.

File: src/shared/dlfcn-util.c (hunk 6 above)
Comment (keszybz) [reply src/shared/dlfcn-util.c]: Makes sense.

File: src/shared/tpm2-util.c (hunk 1 above)
Comment (keszybz) [reply src/shared/tpm2-util.c]: I wanted to keep the same pattern as in the two invocations above… but I guess it doesn't matter, and one line less is better. Will do.

**Review by keszybz — COMMENTED** [review]
Updated with the suggested changed.

> Do you want to this to be included in v249? or later?

v249. It's an important optimization and I think the risk is very low: it's essentially moving a few functions from one place to another. If it compiles (with various configuration option combinations), it almost certainly OK.

> I wonder if the libselinux' gcc constructor idiocy will fuck with us here. i.e. they do complicated setup from a gcc constructor, and we'll possibly delay running that quite a bit, this way, no? previously we ran this at process start for most cases, but now it will be done on dlopen() time. Which hopefully isn't an issue, but might need some careful checking, after all we run things during earlies boot where careful scheduling of the steps matters since so little is set up yet.

No, no. This patchset does not change anything about loading. For our binaries, `libselinux` is still linked statically. The difference is that it's linked from `libsystemd-shared-nnn.so`, and not from `libsystemd.so`. The difference could be for *foreign* binaries, now they are not linked to `libselinux` suddenly. But I think this should be considered a bugfix. This linking was unexpected.

I'll restore the green label from @bluca, since all comments have been addresses and this should be pretty safe.

**Comment by doverride** [comment]
> The difference could be for _foreign_ binaries, now they are not linked to `libselinux` suddenly. But I think this should be considered a bugfix. This linking was unexpected.
> 

Makes me wonder though whether other libraries are linked unexpectedly because as of late quite a bit of access is implied in a lot of places but it might not just be libsystemd pulling in other libraries.

For example linking with libgcrypt causes reading of /etc/gcrypt so all of a sudden many processes started doing stuff like that, There are many more such patterns.

File: src/shared/meson.build [hunk 7]
```diff
@@ -142,10 +142,10 @@ shared_sources = files('''
         install-printf.h
         install.c
         install.h
-        ipvlan-util.c
-        ipvlan-util.h
         ip-protocol-list.c
         ip-protocol-list.h
+        ipvlan-util.c
+        ipvlan-util.h
```
Comment (poettering) [inline src/shared/meson.build:160]: thank god we all use emacs, right? ;-)

**Pull request merged**